### PR TITLE
Fix use scope and highlighting for declarations (Issue #235)

### PR DESCRIPTION
**Pull request by eliasku**
This PR will fix highlighting issue #235 and will fix renaming from parameter or local var/function declarations.

- Fixed usage scope for variables/functions/parameters declaration psi elements.
- Highlighting from usage to declaration fixed
- Renaming from declaration parameter fixed (test case added)
- Unit test added for renaming which isn't work on previous implementation
- Release notes updated

**Before:**
image
image

**After:**
image
image

### CODE REVIEW

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java [hunk 1]
```diff
@@ -62,25 +64,16 @@ public String getName() {
   }
 
   @Override
-  @NotNull
+  @Nullable
   public HaxeComponentName getComponentName() {
-    return new HaxeComponentNameImpl(getNode());
-    ////this.getName();
-    //return findNotNullChildByClass(HaxeComponentName.class);
+    return null;
```
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: Why this change?

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java [hunk 2]
```diff
@@ -213,4 +206,19 @@ public boolean hasModifierProperty(@HaxePsiModifier.ModifierConstant @NonNls @No
...
   }
 
+  @NotNull
+  @Override
+  public SearchScope getUseScope() {
+    final PsiElement parent = getParent();
+    if(parent != null) {
+      if (this instanceof HaxeLocalVarDeclarationPart) {
+        // parent of local var declaration-Part
+        return new LocalSearchScope(parent.getParent());
```
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: I'm growing concerned about code like this.  If we have to expand the tree for some reason, then code that expects a parent of a given type at a certain place, then code like this will break.  And I know I've been taking (and probably writing) code like this a lot.  I think that perhaps we should be using UsefulPsiTreeUtil.getParentOfType() instead of expecting a specific element's tree position.  

What do you think?

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java (hunk 1 above)
Comment (eliasku) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: There was NullPointerException from created HaxeComponentNameImpl, because the passed AST-Node had wrong structure sometimes.

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: So, the answer is to strip the implementation?  The function is obviously used (or you wouldn't have gotten the NPE), so some other functionality depends upon it.

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java (hunk 2 above)
Comment (eliasku) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: Agreed. Fixed.

And code with manual children traversing could be broken for new psi structure (like we saw in getUnderlyingType method before). I will focus more attention to it by the next coding and reviewing

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java (hunk 1 above)
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: There are 41 usages of this function.  I think we still need an implementation.  Was it getNode() that was returning a bad (or unexpected) node?  When/where did the NPE show up?

Comment (eliasku) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: getNode() was right, and NameComponentImpl was created, and then NameComponentImpl::getIdentifier - `findNotNullChildByClass(HaxeIdentifier.class)` throws exception because of `findNotNull__` guarding. I don't remember the case exactly. I could spend a little amount of time to reproduce this case.

NPE had beed shown up for VarDeclaration, not VarPartDeclaration. Notice that getComponentName is overridden for all Implementations, except of VarDeclaration(part is overridden). So we've guarded `getNameIdentifier` which will be Null for VarDeclaration (because the component name inside VarPart).

Comment (eliasku) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: Now, I'm searching the way to reproduce NPE in previous implementation.

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: How about something like this:

PsiIdentifier id = getNameIdentifier();
return null != id ? new HaxeComponenetNameImp(id.getNode()) : null;

Comment (eliasku) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: To reproduce - just type `var a = 0;` and move caret back `var a = <caret>0;` and for LocalVarDeclaration getIdentifier will be required from getComponentName.
```
PsiIdentifier id = getNameIdentifier();
return null != id ? new HaxeComponenetNameImp(id.getNode()) : null;
```
Will cause endless recursion. Notice that `getNameIdentifier` is declared for Mixin

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: > Will cause endless recursion. 

Using your new code, yes.  I was browsing my sources, not the change list, so I got confused.  However, now that the getNameIdentifier() calls getComponentName(), and getComponentName() always returns null, we've stripped the implementation out of *two* functions.

> Notice that getNameIdentifier is declared for Mixin

I don't see a mixin class anywhere in the superclass hierarchy.  Are you saying that *all* subclasses implement it?  If we are going to depend upon that, then we should just remove the method from this (abstract) class; then we'll enforce it.

**Comment by eliasku** [comment]
1. I've reverted `getComponentName` to your suggestion to create `NameComponentImpl` in case if `getNameIdentifier` returns not-null. (Purpose is to **Fix NPE**)

2. I've fixed the `getNameIdentifier` from previous logic to new one. Previous logic was breaking usages to local var declaration referencing. (Purpose is to **Fix highlighting issue**)

3. I've made typed searching. Block-statement type parent for LocalSearchScope creation, but only in case if we're in context of Local-Declaration-Node. (I'm not sure 100% how IDEA works with `getUseScope`, because previously I thought that IDEA will not search in comments outside of LocalSearchScope while refactoring. It was wrong. But at least it fixes the renaming of same names  references outside )

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java [hunk 3]
```diff
@@ -213,4 +208,16 @@ public boolean hasModifierProperty(@HaxePsiModifier.ModifierConstant @NonNls @No
     return this.getModifierList().hasModifierProperty(name);
   }
 
+  @NotNull
+  @Override
+  public SearchScope getUseScope() {
+    final PsiElement localVar = UsefulPsiTreeUtil.getParentOfType(this, HaxeLocalVarDeclaration.class);
+    if(localVar != null) {
+      final PsiElement outerBlock = UsefulPsiTreeUtil.getParentOfType(localVar, HaxeBlockStatement.class);
```
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: So, does it work in this case?:

```
class bar {
  function doSomething();
}
class foo {
  var a:bar = new bar; 
  function printNtimes(msg:String, n:Int) :Void for( a in 0...n ) Sys.println( a.<caret>_toString() + ":" + msg);
  //                                                                              ^^^
}
```

Comment (eliasku) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: It doesn't work because `a` inside `for` is not local-variable-declaration. For `for` statement I've added support for renaming from using to declaration. It was missing NameIdentifier. Look new commit, your example will work correctly.

**Comment by eliasku** [comment]
Thank you, @EBatTiVo! `for` statement declaration is also fixed now!
Before:
image
After:
image

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java (hunk 3 above)
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: That is a good fix.  However, the point of the sample was really to point out that local scopes (such as functions) aren't *always* surrounded by HaxeBlockStatements.  In my version of the example, the variable 'a' inside of the for loop hides the variable with the same name and different type declared at the class level.  Therefore, locating the scope requires searching for more than an enclosing block statement; it requires looking for enclosing function definitions, too.

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: I guess I missed this statement you made.

> It doesn't work because a inside for is not local-variable-declaration.

If a variable being declared on a for statement is not a local variable declaration, what is it?  More importantly: Would somebody writing code expect searching, highlighting, refactoring, etc. to act as if it *were* a local variable?

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: And then, after making these arguments, I go look at your unit test more closely and it's working as I expect.

**Comment by EBatTiVo** [comment]
@eliasku Would you mind looking at #136, which is very closely related, and see if this change fixed that, too?

**Comment by eliasku** [comment]
@EBatTiVo I think so. I reproduce this issue #136 it in `0.9.9`. For actual `origin/master` there is no issue anymore.

**Comment by EBatTiVo** [comment]
That wording is a bit confusing.  I take it to mean that you tested and the issue is fixed.  Correct?

**Comment by eliasku** [comment]
That's right. I've corrected the previous comment.
